services/venice-controller/src/main/java/com/linkedin/venice/controller/server/CreateVersion.java

```diff
@@ -71,6 +74,356 @@ public CreateVersion(
     this.disableParentRequestTopicForStreamPushes = disableParentRequestTopicForStreamPushes;
   }
 
+  static void extractOptionalParamsFromRequestTopicRequest(
+      Request httpRequest,
+      RequestTopicForPushRequest request,
+      boolean isAclEnabled) {
+    request.setPartitioners(httpRequest.queryParamOrDefault(PARTITIONERS, null));
+
+    request.setSendStartOfPush(
+        Utils.parseBooleanFromString(httpRequest.queryParamOrDefault(SEND_START_OF_PUSH, "false"), SEND_START_OF_PUSH));
+
+    request.setSorted(
+        Utils.parseBooleanFromString(
+            httpRequest.queryParamOrDefault(PUSH_IN_SORTED_ORDER, "false"),
+            PUSH_IN_SORTED_ORDER));
+
+    request.setWriteComputeEnabled(
+        Utils.parseBooleanFromString(
+            httpRequest.queryParamOrDefault(IS_WRITE_COMPUTE_ENABLED, "false"),
+            IS_WRITE_COMPUTE_ENABLED));
+
+    /*
+     * Version-level rewind time override, and it is only valid for hybrid stores.
+     */
+    request.setRewindTimeInSecondsOverride(
+        Long.parseLong(httpRequest.queryParamOrDefault(REWIND_TIME_IN_SECONDS_OVERRIDE, "-1")));
+
+    /*
+     * Version level override to defer marking this new version to the serving version post push completion.
+     */
+    request.setDeferVersionSwap(
+        Utils.parseBooleanFromString(httpRequest.queryParamOrDefault(DEFER_VERSION_SWAP, "false"), DEFER_VERSION_SWAP));
+
+    request.setTargetedRegions(httpRequest.queryParamOrDefault(TARGETED_REGIONS, null));
+
+    request.setRepushSourceVersion(Integer.parseInt(httpRequest.queryParamOrDefault(REPUSH_SOURCE_VERSION, "-1")));
+
+    request.setSourceGridFabric(httpRequest.queryParamOrDefault(SOURCE_GRID_FABRIC, null));
+
+    request.setCompressionDictionary(httpRequest.queryParamOrDefault(COMPRESSION_DICTIONARY, null));
+
+    // Retrieve certificate from request if ACL is enabled
+    request.setCertificateInRequest(isAclEnabled ? getCertificate(httpRequest) : null);
+  }
+
+  private static void verifyPartitioner(PartitionerConfig storePartitionerConfig, Set<String> partitionersFromRequest) {
+    // If partitioners are provided, check if the store partitioner is in the list
+    if (partitionersFromRequest != null && !partitionersFromRequest.isEmpty()
```

Comment:
is the isEmpty test here necessary? If a set is empty, your third test contains(xXX) will return false.

Reply:
Yes, the `partitionersFromRequest.isEmpty()` check is necessary. The intent of the method is to verify if the optional partitioner list provided by the user contains the partitioner from the store’s configuration. If we do not explicitly check for isEmpty, and the user provides an empty set, we would inadvertently throw an exception. This is not the expected behavior because an empty set implies that no specific partitioners were requested, and we should simply fall back to the store’s partitioner without raising an error.